envoy: Generalize access logging.

jrajahalme opened the pull request:
Define extensible layer 7 `oneof` for accesslog protobuf.  Deprecate
the old http fields in the main access log protobuf
message and convert to new format when received.

This change makes it clearer how to extend access logging for new L7
protocols while maintaining backwards compatibility with existing
pods.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/5219)
<!-- Reviewable:end -->

rlenglet commented on envoy/cilium/accesslog.proto:
```diff
@@ -20,11 +20,30 @@ enum EntryType {
 }
 
 message HttpLogEntry {
+  Protocol http_protocol = 1;
```
Rename `Protocol` -> `HttpProtocol`.

jrajahalme replied:
OK.

rlenglet commented on envoy/cilium/accesslog.proto:
```diff
@@ -20,11 +20,30 @@ enum EntryType {
+  uint32 status = 7;      // Envoy ":status" header, zero for request
+}
+
+// Deprecated old message format, to be removed in a future Cilium release
```
I wouldn't say that this is deprecated. :)
I would only say that some of the fields documented as deprecated will be removed in a future Cilium release.

jrajahalme replied:
The comment was a remnant of an earlier work-in-progress version, thanks for spotting it!

rlenglet commented on envoy/cilium/accesslog.proto:
```diff
@@ -34,27 +53,32 @@ message HttpLogEntry {
+  //
+  // Deprecated HTTP fields
+  //
+  Protocol http_protocol = 2;
```
Add the `deprecated` option to every field here, cf. https://developers.google.com/protocol-buffers/docs/proto3, e.g.:
```
HttpProtocol http_protocol = 2 [deprecated=true];
```

jrajahalme replied:
OK, thanks for the pointer!

rlenglet commented on envoy/cilium/accesslog.proto:
```diff
@@ -34,27 +53,32 @@ message HttpLogEntry {
+  }
+
+  //
+  // Deprecated HTTP fields
```
```
// Deprecated HTTP fields. Use the http field above instead.
```

jrajahalme replied:
OK.

rlenglet reviewed (approved):
Awesome! Thanks!

jrajahalme committed:
envoy: Generalize access logging.

Define extensible layer 7 'oneof' for accesslog protobuf.  Deprecate
old access loggingthe old http fields in the main access log protobuf
message and convert to new format when received.

This change makes it clearer how to extend access logging for new L7
protocols while maintaining backwards compatibility with existing
pods.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>
